Feat/redeem payment command 2850

rosalinekarr opened the pull request:
# Problem

We don't have a simple way for users to redeem vouchers for a given storage deal.

# Solution

Create a `go-filecoin deals redeem` command for redeeming the first valid voucher on a given deal.

Resolves #2850

acruikshank commented:
@rosalinekarr there are changes in this PR that aren't strictly necessary for #2850. Please describe what they are and why they are in this PR.

rosalinekarr commented:
@acruikshank The only other thing I addressed in this PR was moving the `go-filecoin client list-deals` and `go-filecoin miner list-deals` commands into a new `go-filecoin deals list` command. The functionality and tests for these commands are unchanged, just moved. I included these changes because they seemed small and related enough to creating the `go-filecoin deals` command to justify making them.

If you want to ignore these changes, just skip the first two commits to the branch when diffing, or if you'd like, I can remove them and make a new PR for these changes.

rosalinekarr commented:
@acruikshank never mind, I went ahead and extracted it to a new PR here: https://github.com/filecoin-project/go-filecoin/pull/2910

acruikshank commented on porcelain/storagedeals.go:
```diff
@@ -140,3 +142,84 @@ func filterDealChannel(dealCh <-chan *StorageDealLsResult, filterFunc func(*stor
+
+// DealRedeem redeems a voucher for the deal with the given cid and returns
+// either the cid of the created redeem message or an error
+func DealRedeem(ctx context.Context, plumbing dealRedeemPlumbing, fromAddr address.Address, dealCid cid.Cid, gasPrice types.AttoFIL, gasLimit types.GasUnits) (cid.Cid, error) {
```
This plumbing function needs a test to confirm it correctly chooses the best voucher from a deal and sends a properly formatted message.

acruikshank commented on porcelain/storagedeals.go:
```diff
@@ -140,3 +142,84 @@ func filterDealChannel(dealCh <-chan *StorageDealLsResult, filterFunc func(*stor
+
+// DealRedeemPreview previews the redeem method for a deal and returns the
+// expected gas used
+func DealRedeemPreview(ctx context.Context, plumbing dealRedeemPlumbing, fromAddr address.Address, dealCid cid.Cid) (types.GasUnits, error) {
```
This plumbing function needs a test to confirm that it creates the deal redeem params the same way as `DealRedeem` and properly formats a message sent to `MessagePreview`.

acruikshank commented on commands/deals_daemon_test.go:
```diff
@@ -0,0 +1,66 @@
+	newWalletBalance, err := minerDaemon.WalletBalance(ctx, minerAddress)
+	require.NoError(t, err)
+
+	expectedBalanceDiff := types.NewAttoFILFromFIL(0).String()
```
This shouldn't be zero. It should be the value of the voucher.

acruikshank commented on porcelain/storagedeals_test.go:
```diff
@@ -139,3 +140,162 @@ func TestDealMinerLs(t *testing.T) {
+	resultCid, err := porcelain.DealRedeem(context.Background(), plumbing, fromAddr, dealCid, types.NewAttoFILFromFIL(0), types.NewGasUnits(0))
+	require.NoError(t, err)
+
+	assert.Equal(t, messageCid, resultCid)
```
Running the equality assertions in the stub obscures the meaning of this test. It's correct and sufficient, but it takes some unnecessary analysis on the readers part to establish what's under test.
